### PR TITLE
finalize: strip segments that contain only EM_ASM/EM_JS data

### PR DESCRIPTION
**Pull request by sbc100**
If we find a data segment whose entire contents is EM_JS or EM_ASM
strings then strip it from the binary.

This change also introduces a new way to exact EM_ASM strings based
on the __start_ and _stop_ sections symbols.   This code will no be active
until the emscripten-side change lands.

See: https://github.com/emscripten-core/emscripten/pull/13443

### CODE REVIEW

**Comment by sbc100** [comment]
Hmm.. how come the UI is saying I can merge this without a review?   Does binaryen not require an lgtm?

**Comment by tlively** [comment]
That's weird, it should require an LGTM

**Comment by sbc100** [comment]
Ah.. its because I wasn't targeting main.   Fixed.

File: src/wasm/wasm-emscripten.cpp
Comment (tlively) [inline src/wasm/wasm-emscripten.cpp:424]: This would be more robust against side effects if it created a block that evaluated and dropped all the operands of the `MemoryInit`, but perhaps that's overkill considering how much we can assume about the output of wasm-ld here.

File: src/wasm/wasm-emscripten.cpp
Comment (tlively) [inline src/wasm/wasm-emscripten.cpp:462]:
```suggestion
      // Entire segment is contains JS strings.  Remove it.
```

**Review by tlively — APPROVED** [review]

File: src/wasm/wasm-emscripten.cpp
Comment (sbc100) [reply src/wasm/wasm-emscripten.cpp:424]: I see... I wonder if there are other places this is done?  Maybe there is some kind of helper for this already?

Comment (tlively) [reply src/wasm/wasm-emscripten.cpp:424]: I don't know of a general utility for this, but here's an example where this is done: https://github.com/WebAssembly/binaryen/blob/74ccdb87b3390e003b658b847526dad8722083f8/src/passes/MemoryPacking.cpp#L413-L416

`Builder::blockify` makes this simple enough that we probably don't need a separate utility for it (although a dedicated utility couldn't hurt).

**Comment by sbc100** [comment]
This change now handle EM_ASM strings too.

File: src/wasm/wasm-emscripten.cpp
Comment (tlively) [inline src/wasm/wasm-emscripten.cpp:359]: Are planning to remove this fallback code in the future? What is stopping us from removing it now?

File: test/lit/wasm-emscripten-finalize/em_asm.wat
Comment (tlively) [inline test/lit/wasm-emscripten-finalize/em_asm.wat:1]:
```suggestion
;; Test that em_asm string are extracted correctly when the __start_em_asm
```

File: test/lit/wasm-emscripten-finalize/em_js.wat
Comment (tlively) [inline test/lit/wasm-emscripten-finalize/em_js.wat:12]: Would this change correctly test that the entire segment is stripped, rather than just changed?
```suggestion
;; CHECK-NOT: (i32.const 512)
```

File: test/lit/wasm-emscripten-finalize/em_js.wat
Comment (tlively) [inline test/lit/wasm-emscripten-finalize/em_js.wat:16]: Why aren't the JS strings zeroed out?

Comment (sbc100) [reply test/lit/wasm-emscripten-finalize/em_js.wat:16]: I've only done the zeroing out thing for the em_asm path so far.     The reason is that I don't currently use the start/end symbols for the em_js path at all.   I think we can do that a followup perhaps, and maybe even completely remove the current 1-export-per-em-js thing.

File: src/wasm/wasm-emscripten.cpp
Comment (sbc100) [reply src/wasm/wasm-emscripten.cpp:359]: Yes.   Mostly its a transition thing.    I can't land the emscripten-side change first because it actually changes the way that arg0 gets passed the em_asm functions.   llvm ends up storing that address of the string in memory and then loading it back (at least in -O0) which is not currently supported by the AsmConstWalker.   So if I land the emscripten side first it would break this pass.

File: test/lit/wasm-emscripten-finalize/em_js.wat
Comment (tlively) [reply test/lit/wasm-emscripten-finalize/em_js.wat:16]: Gotcha. The difference between EM_JS and EM_ASM still confuses me, but that's ok.

**Review by tlively — APPROVED** [review]

**Pull request merged**
